src/deepsparse/loggers/prometheus_logger.py

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implementation of the Prometheus Logger
+"""
+import logging
+import os
+from collections import defaultdict
+from typing import Any, Optional
+
+from deepsparse.loggers import BaseLogger, MetricCategories
+from prometheus_client import REGISTRY, Summary, start_http_server, write_to_textfile
+
+
+__all__ = ["PrometheusLogger"]
+
+_LOGGER = logging.getLogger(__name__)
+SUPPORTED_DATA_TYPES = [int, float]
+
+
+class PrometheusLogger(BaseLogger):
+    """
+    DeepSparse logger that continuously exposes the collected logs over the
+    Prometheus python client at the specified port.
+
+    :param port: the port used by the client. Default is 6100
+    :param text_log_save_frequency: the frequency of saving the text log
+        files. E.g. if `text_log_save_frequency` = 10, text logs are dumped
+        after every tenth forward pass. Default set to 10
+    :param text_log_save_dir: the directory where the text log files
+        are saved. By default, the python working directory
+    :param text_log_file_name: the name of the text log file.
+        Default: `prometheus_logs.prom`
+    """
+
+    def __init__(
+        self,
+        port: int = 6100,
+        text_log_save_frequency: int = 10,
+        text_log_save_dir: str = os.getcwd(),
+        text_log_file_name: Optional[str] = None,
+    ):
+
+        self.port = port
+        # until we have done research into persistent
+        # logs, lets save logs as .prom files
+        self.text_log_save_frequency = text_log_save_frequency
+        self.text_log_file_path = os.path.join(
+            text_log_save_dir, text_log_file_name or "prometheus_logs.prom"
+        )
+        self._prometheus_metrics = defaultdict(str)
+
+        self._setup_client()
+        self._counter = 0
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        Collect information from the pipeline and pipe it them to the stdout
+
+        :param identifier: The name of the thing that is being logged.
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+
+        # needs to adhere to
+        # https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+        formatted_identifier = identifier.replace(".", "__")
```

Comment:
I am not paying too much focus to the best practices of Prometheus logger, but will need to look into it later: https://prometheus.io/docs/practices/naming/

Reply:
also need to replace any whitespaces. why double underscore?

Reply:
Our agreed-upon structure of an identifier is:
`<some_string_1>.<some_string_2>.<some_string_3>...`
However, as stated here: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

```
The metric name specifies the general feature of a system that is measured (e.g. http_requests_total - the total number of HTTP requests received). It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex [a-zA-Z_:][a-zA-Z0-9_:]*.
```
So the non-letter chars that may be used by Prometheus are `:` and `_` right? Well not exactly:
```
The colons are reserved for user-defined recording rules. They should not be used by exporters or direct instrumentation.
```

So we are left only with the underscores.

Usually, a Prometheus metric looks like this: 
`total_time_elapsed_seconds`

But the identifiers can look like `pipeline_input.something1.something2`. If we just replaced all the dots with underscores we end up with `pipeline_inputs_something1_something2` - we lose the tight relationship (e.g. when filtering) between the string `pipeline` and  string `inputs`. So I decided that the least problematic formatting would be `pipeline_inputs__something1__something2`. It's not great, but it's clear enough.

Note: `Label names beginning with __ are reserved for internal use.`

Reply:
yup, makes sense - only concern would be downstream users potentially missing the double